core/src/main/java/org/elasticsearch/index/query/functionscore/DecayFunctionBuilder.java

```diff
@@ -19,25 +19,56 @@
 
 package org.elasticsearch.index.query.functionscore;
 
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Explanation;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.geo.GeoDistance;
+import org.elasticsearch.common.geo.GeoPoint;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.lucene.search.function.CombineFunction;
+import org.elasticsearch.common.lucene.search.function.LeafScoreFunction;
+import org.elasticsearch.common.lucene.search.function.ScoreFunction;
+import org.elasticsearch.common.unit.DistanceUnit;
+import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.fielddata.*;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.core.DateFieldMapper;
+import org.elasticsearch.index.mapper.core.NumberFieldMapper;
+import org.elasticsearch.index.mapper.geo.GeoPointFieldMapper;
+import org.elasticsearch.index.query.QueryParseContext;
+import org.elasticsearch.index.query.QueryValidationException;
 import org.elasticsearch.search.MultiValueMode;
 
 import java.io.IOException;
 import java.util.Locale;
+import java.util.Objects;
 
-public abstract class DecayFunctionBuilder extends ScoreFunctionBuilder {
+public abstract class DecayFunctionBuilder extends ScoreFunctionBuilder<DecayFunctionBuilder> {
 
     protected static final String ORIGIN = "origin";
     protected static final String SCALE = "scale";
     protected static final String DECAY = "decay";
     protected static final String OFFSET = "offset";
 
+    public static final double DEFAULT_DECAY = 0.5;
+    public static final double DEFAULT_OFFSET_NUMBER = 0.0d;
+    public static final String DEFAULT_OFFSET_GEO = "0km";
+    public static final String DEFAULT_OFFSET_DATE = "0d";
+    public static final MultiValueMode DEFAULT_MULTI_VALUE_MODE = MultiValueMode.MIN;
+
     private String fieldName;
     private Object origin;
     private Object scale;
-    private double decay = -1;
-    private Object offset;
-    private MultiValueMode multiValueMode = null;
+    private double decay = DEFAULT_DECAY;
+    private Object offset;  // field type specific default not known in advance
+    private MultiValueMode multiValueMode = DEFAULT_MULTI_VALUE_MODE;
+
+    /**
+     * Override this function if you want to produce your own scorer.
+     */
+    public abstract DecayFunction getDecayFunction();
```

Comment:
comment is conditional but the method is abstract. it's not that you can choose whether to override it or not...